File: step_06_linked_list/6.3_medium_problems/P002_reverse_a_linked_list/solution.py
"""
Problem: Reverse a Linked List
Difficulty: EASY | XP: 10
LeetCode #206

Reverse a singly linked list iteratively and recursively.
"""
from typing import Optional, List


# ============================================================
# NODE DEFINITION
# ============================================================
class ListNode:
    def __init__(self, val: int = 0, next: 'ListNode' = None):
        self.val = val
        self.next = next


# ============================================================
# APPROACH 1: BRUTE FORCE -- USING ARRAY
# Time: O(n)  |  Space: O(n)
# Store values, overwrite in reverse. Changes values, not pointers.
# ============================================================
def reverse_brute_force(head: Optional[ListNode]) -> Optional[ListNode]:
    """Reverse by collecting values and overwriting."""
    if not head:
        return None

    # Collect values
    vals = []
    temp = head
    while temp:
        vals.append(temp.val)
        temp = temp.next

    # Overwrite in reverse
    temp = head
    for val in reversed(vals):
        temp.val = val
        temp = temp.next

    return head


# ============================================================
# APPROACH 2: ITERATIVE (OPTIMAL)
# Time: O(n)  |  Space: O(1)
# Three-pointer technique: prev, curr, next.
# ============================================================
def reverse_iterative(head: Optional[ListNode]) -> Optional[ListNode]:
    """Reverse by flipping pointers one by one."""
    prev = None
    curr = head

    while curr:
        nxt = curr.next   # save next
        curr.next = prev  # reverse arrow
        prev = curr       # advance prev
        curr = nxt        # advance curr

    return prev  # new head


# ============================================================
# APPROACH 3: RECURSIVE
# Time: O(n)  |  Space: O(n) call stack
# Reverse the rest, then fix the backward link.
# ============================================================
def reverse_recursive(head: Optional[ListNode]) -> Optional[ListNode]:
    """Reverse recursively -- base case propagates new head."""
    # Base case: empty or single node
    if not head or not head.next:
        return head

    # Reverse everything after head
    new_head = reverse_recursive(head.next)

    # head.next is now tail of reversed sublist; point it back
    head.next.next = head
    head.next = None  # head becomes the new tail

    return new_head


# ============================================================
# UTILITY: Build list from array
# ============================================================
def build_from_list(arr: List[int]) -> Optional[ListNode]:
    if not arr:
        return None
    head = ListNode(arr[0])
    current = head
    for val in arr[1:]:
        current.next = ListNode(val)
        current = current.next
    return head


# ============================================================
# UTILITY: Print list
# ============================================================
def print_list(head: Optional[ListNode]) -> str:
    parts = []
    current = head
    while current:
        parts.append(str(current.val))
        current = current.next
    return " -> ".join(parts) if parts else "(empty)"


if __name__ == "__main__":
    print("=== Reverse a Linked List ===\n")

    # Approach 1: Brute force
    head1 = build_from_list([1, 2, 3, 4, 5])
    print(f"Brute force: {print_list(reverse_brute_force(head1))}")

    # Approach 2: Iterative
    head2 = build_from_list([1, 2, 3, 4, 5])
    print(f"Iterative:   {print_list(reverse_iterative(head2))}")

    # Approach 3: Recursive
    head3 = build_from_list([1, 2, 3, 4, 5])
    print(f"Recursive:   {print_list(reverse_recursive(head3))}")

    # Edge cases
    print("\n--- Edge Cases ---")
    print(f"Empty:       {print_list(reverse_iterative(None))}")
    print(f"Single node: {print_list(reverse_iterative(build_from_list([42])))}")
    print(f"Two nodes:   {print_list(reverse_iterative(build_from_list([1, 2])))}")

File: step_06_linked_list/6.3_medium_problems/P002_reverse_a_linked_list/Solution.java
/**
 * Problem: Reverse a Linked List
 * Difficulty: EASY | XP: 10
 * LeetCode #206
 *
 * Reverse a singly linked list iteratively and recursively.
 *
 * @author DSA_Nova
 */
public class Solution {

    // ============================================================
    // NODE DEFINITION
    // ============================================================
    static class ListNode {
        int val;
        ListNode next;

        ListNode(int val) {
            this.val = val;
            this.next = null;
        }
    }

    // ============================================================
    // APPROACH 1: BRUTE FORCE -- USING ARRAY
    // Time: O(n)  |  Space: O(n)
    // ============================================================
    /**
     * Store values in array, overwrite node values in reverse order.
     * Note: Changes values, not pointers. Works when pointer identity
     * doesn't matter.
     */
    public static ListNode reverseBruteForce(ListNode head) {
        if (head == null) return null;

        // Count nodes
        int count = 0;
        ListNode temp = head;
        while (temp != null) {
            count++;
            temp = temp.next;
        }

        // Store values
        int[] vals = new int[count];
        temp = head;
        for (int i = 0; i < count; i++) {
            vals[i] = temp.val;
            temp = temp.next;
        }

        // Overwrite in reverse
        temp = head;
        for (int i = count - 1; i >= 0; i--) {
            temp.val = vals[i];
            temp = temp.next;
        }

        return head;
    }

    // ============================================================
    // APPROACH 2: ITERATIVE (OPTIMAL)
    // Time: O(n)  |  Space: O(1)
    // ============================================================
    /**
     * Three-pointer technique: prev, curr, next.
     * Reverse each next pointer one by one in a single pass.
     */
    public static ListNode reverseIterative(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;

        while (curr != null) {
            ListNode next = curr.next; // save next
            curr.next = prev;          // reverse arrow
            prev = curr;               // advance prev
            curr = next;               // advance curr
        }

        return prev; // prev is the new head
    }

    // ============================================================
    // APPROACH 3: RECURSIVE
    // Time: O(n)  |  Space: O(n) call stack
    // ============================================================
    /**
     * Recursively reverse the rest of the list, then fix the
     * backward link from head.next back to head.
     */
    public static ListNode reverseRecursive(ListNode head) {
        // Base case: empty list or single node
        if (head == null || head.next == null) {
            return head;
        }

        // Recursively reverse everything after head
        ListNode newHead = reverseRecursive(head.next);

        // head.next is now the tail of the reversed sublist
        // Make it point back to head
        head.next.next = head;
        head.next = null; // head is now the tail

        return newHead; // propagate new head up
    }

    // ============================================================
    // UTILITY: Build list from array
    // ============================================================
    public static ListNode buildFromArray(int[] arr) {
        if (arr == null || arr.length == 0) return null;
        ListNode head = new ListNode(arr[0]);
        ListNode current = head;
        for (int i = 1; i < arr.length; i++) {
            current.next = new ListNode(arr[i]);
            current = current.next;
        }
        return head;
    }

    // ============================================================
    // UTILITY: Print list
    // ============================================================
    public static void printList(ListNode head) {
        StringBuilder sb = new StringBuilder();
        ListNode current = head;
        while (current != null) {
            sb.append(current.val);
            if (current.next != null) sb.append(" -> ");
            current = current.next;
        }
        System.out.println(sb.length() > 0 ? sb.toString() : "(empty)");
    }

    public static void main(String[] args) {
        System.out.println("=== Reverse a Linked List ===\n");

        // Approach 1: Brute force (value swap)
        System.out.println("Approach 1 - Brute Force (array):");
        ListNode head1 = buildFromArray(new int[]{1, 2, 3, 4, 5});
        System.out.print("Before: "); printList(head1);
        head1 = reverseBruteForce(head1);
        System.out.print("After:  "); printList(head1);

        // Approach 2: Iterative
        System.out.println("\nApproach 2 - Iterative:");
        ListNode head2 = buildFromArray(new int[]{1, 2, 3, 4, 5});
        System.out.print("Before: "); printList(head2);
        head2 = reverseIterative(head2);
        System.out.print("After:  "); printList(head2);

        // Approach 3: Recursive
        System.out.println("\nApproach 3 - Recursive:");
        ListNode head3 = buildFromArray(new int[]{1, 2, 3, 4, 5});
        System.out.print("Before: "); printList(head3);
        head3 = reverseRecursive(head3);
        System.out.print("After:  "); printList(head3);

        // Edge cases
        System.out.println("\n--- Edge Cases ---");

        System.out.print("Empty list: ");
        printList(reverseIterative(null));

        System.out.print("Single node: ");
        printList(reverseIterative(buildFromArray(new int[]{42})));

        System.out.print("Two nodes: ");
        printList(reverseIterative(buildFromArray(new int[]{1, 2})));
    }
}
